server/channels/app/platform/service.go

```diff
@@ -202,19 +175,73 @@ func New(sc ServiceConfig, options ...Option) (*PlatformService, error) {
 	searchEngine.RegisterBleveEngine(bleveEngine)
 	ps.SearchEngine = searchEngine
 
-	// Step 4: Init Enterprise
-	// Depends on step 3 (s.SearchEngine must be non-nil)
+	// Step 3: Init Enterprise
+	// Depends on step 2 (s.SearchEngine must be non-nil)
 	ps.initEnterprise()
 
-	// Step 5: Init Metrics
+	// Step 4: Init Metrics
 	if metricsInterfaceFn != nil && ps.metricsIFace == nil { // if the metrics interface is set by options, do not override it
 		ps.metricsIFace = metricsInterfaceFn(ps, *ps.configStore.Get().SqlSettings.DriverName, *ps.configStore.Get().SqlSettings.DataSource)
 	}
 
+	// Step 5: Create the sqlstore. This is moved out of the
+	// ps.newStore function because we need to load the license without loading
+	// the store entirely, because the full store includes the localCacheLayer.
+	// And the localCacheLayer needs license to be loaded first to allow Redis or not.
+	var err error
+	ps.sqlStore, err = sqlstore.New(ps.Config().SqlSettings, ps.Log(), ps.metricsIFace)
+	if err != nil {
+		return nil, err
+	}
+
+	// Step 6: Load license.
+	// Store needs to be loaded before this.
+	if model.BuildEnterpriseReady == "true" {
+		ps.LoadLicense()
+	}
+
+	license := ps.License()
+	// We need to update the store with the license.
+	// This is to prevent search replicas being used without a license.
+	ps.sqlStore.UpdateLicense(license)
+
+	// Step 7: Cache provider.
+	cacheConfig := ps.configStore.Get().CacheSettings
+	if *cacheConfig.CacheType == model.CacheTypeLRU {
+		ps.cacheProvider = cache.NewProvider()
+	} else if *cacheConfig.CacheType == model.CacheTypeRedis {
+		// Only allow Redis in an HA setting.
+		if license != nil && *license.Features.Cluster {
+			ps.cacheProvider, err = cache.NewRedisProvider(
+				&cache.RedisOptions{
+					RedisAddr:     *cacheConfig.RedisAddress,
+					RedisPassword: *cacheConfig.RedisPassword,
+					RedisDB:       *cacheConfig.RedisDB,
+					DisableCache:  *cacheConfig.DisableClientCache,
+				},
+			)
+		} else {
+			ps.Log().Warn("No license found, or license does not have clustering feature. Falling back to LRU cache")
+			ps.UpdateConfig(func(cfg *model.Config) {
+				*cfg.CacheSettings.CacheType = model.CacheTypeLRU
+			})
+			cacheConfig = ps.configStore.Get().CacheSettings
+			ps.cacheProvider = cache.NewProvider()
+		}
+	}
```

Comment:
If CacheType is LRU, we simply do `ps.cacheProvider = cache.NewProvider()`, but if CacheType is Redis and the license does not support it, we update the config and recompute cacheConfig before doing `ps.cacheProvider = cache.NewProvider()`. Why that difference?

Reply:
Because later the cacheConfig is logged, to identify what cache backend is being used. The order of L228 and 229 doesn't matter if that's what you are asking.